nym-vpn-desktop/src-tauri/src/commands/node_location.rs

```diff
@@ -85,38 +84,152 @@ pub async fn get_node_location(
+            let mut name = location.country_name;
+            // TODO yes this is what we get from the API for UK
+            // let's use something more friendly
+            if name == "United Kingdom of Great Britain and Northern Ireland" {
+                name = "United Kingdom".to_string();
+            }
+
+            Country {
+                name,
+                code: location.two_letter_iso_country_code,
+            }
+        })
+        // sort countries by name
+        .sorted_by(|a, b| a.name.cmp(&b.name))
+        .collect::<Vec<_>>();
+
+    debug!("fetched countries count [{}]", list.len());
+    trace!("fetched countries {list:#?}");
+
+    Ok(list)
+}
+
+#[instrument(skip_all)]
+pub async fn get_exit_countries() -> Result<Vec<Country>, CmdError> {
+    debug!("get_exit_countries");
+
+    // future::join_all will collects the responses in the same
+    // order
+    let urls = vec![explorer_api::get_url(), nym_api::get_url()];
+
+    debug!("fetching countries from Explorer and Nym api");
+    // concurrently fetch both explorer and nym APIs
+    let responses = future::join_all(urls.into_iter().map(|url| {
+        let client = &HTTP_CLIENT;
+        async move {
+            client.get(&url).send().await.map_err(|e| {
+                error!("HTTP request GET {url} failed: {e}");
+                HttpError::RequestError(e.status())
+            })
+        }
+    }))
+    .await;
+    trace!("fetching done");
+
+    // filter out any failed requests
+    let mut responses = responses
+        .into_iter()
+        .filter_map(|res| res.ok())
+        .collect::<Vec<_>>();
+
+    // if one of the requests failed, return an error
+    if responses.len() != 2 {
+        return Err(CmdError::new(
+            CmdErrorSource::InternalError,
+            "failed to fetch node locations".to_string(),
+        ));
+    }
+    let explorer_response = responses.remove(0);
+    let nym_api_response = responses.remove(0);
+
+    debug!("deserializing json responses");
+    let explorer_json = explorer_api::deserialize_json(explorer_response)
+        .await
+        .map_err(|_| {
+            CmdError::new(
+                CmdErrorSource::InternalError,
+                "failed to fetch node locations".to_string(),
+            )
+        })?;
+
+    let nym_api_json = nym_api::deserialize_json(nym_api_response)
+        .await
+        .map_err(|_| {
+            CmdError::new(
+                CmdErrorSource::InternalError,
+                "failed to fetch node locations".to_string(),
+            )
+        })?;
+
+    debug!("parsing responses");
+    let list = nym_api_json
+        .into_iter()
+        // only retain gateways with IP packet router
+        // mapping to a list of identity keys
+        .filter_map(|gateway| {
+            gateway
+                .self_described
+                .and_then(|desc| desc.ip_packet_router)
+                .map(|_| gateway.bond.gateway().identity_key.clone())
```

Comment:
I think this works, but it's a bit hard to read. Maybe just do another `filter_map` in the main iterator, and append the identity key to a tuple, so that it gets carried forward where you need it.

Reply:
sorry for this one I dont see what you mean, the code is valid (yes it works), so unless you share some code I'm not able to change it